### PR TITLE
Add optional offline mirror pruning

### PR DESCRIPTION
**Pull request by dguo**
**Summary**

This change is an implementation of yarnpkg/rfcs#49, which stems from #2109.

**Test plan**

I tested the feature locally with two scenarios where `.yarnrc` has `yarn-offline-mirror-pruning` set to `true`:

* Upgrading a package:
    * `$ yarn add is-positive-integer@1.0.0`
    * Check that the offline mirror has the `is-positive-integer` tarball, along with all its dependencies
    * `$ yarn upgrade is-positive-integer@1.1.1`
    * Check that the offline mirror only has the new version of `is-positive-integer`, which doesn't have any dependencies
* Removing a package:
    * `$ yarn add left-pad right-pad`
    * Check that the offline mirror has both the `left-pad` and `right-pad` tarballs
    * `$ yarn remove left-pad`
    * Check that the offline mirror only has `right-pad`

I also added two Jest test cases, one for upgrade and one for remove. They verify that sub-dependencies are also pruned.

For documentation, I'd be happy to submit a pull request for that as well, but I don't see any for the offline mirror feature on the main Yarn website. I've just been referencing [this blog post](https://yarnpkg.com/blog/2016/11/24/offline-mirror/).

### CODE REVIEW

File: src/cli/commands/install.js [hunk 1]
```diff
@@ -572,14 +572,44 @@ export class Install {
...
+    const mirror = this.config.getOfflineMirrorPath();
+    if (!mirror) {
+      return;
+    }
+
+    const requiredTarballs = new Set();
+    for (const dependency in lockfile) {
+      const resolved = lockfile[dependency].resolved;
+      if (resolved) {
+        requiredTarballs.add(resolved.split('#')[0]);
```
Comment (dguo) [inline src/cli/commands/install.js]: I'm not sure if it is okay to depend on the format of `resolved` like this.

Comment (bestander) [inline src/cli/commands/install.js]: The tests would fail if there are breaking changes.
I suppose it should be fine.

**Comment by bestander** [comment]
Looks nice, how about a test for add command as well?

**Comment by bestander** [comment]
Tests seem to fail for some other reason

**Comment by dguo** [comment]
Yep, I considered adding a test for add as well, but I couldn't come up with a situation where pruning would apply. Maybe there's some uncommon semver combination that I haven't thought of? If not, should I just copy over the upgrade test fixture and use that? So it would simulate:
`$ yarn add dep-a@1.0.0`
`$ yarn add dep-a@1.1.0`

The tests fail for me locally as well, but the last I tried, the same tests failed on the master branch. I attributed the failures to flakiness, but would you like me to dig into them?

**Comment by bestander** [comment]
I think you are right, `add` that would cause package remove is only when a package already exists and gets updated.
RE tests, seems like it is flakiness

**Comment by bestander** [comment]
@dguo, could you rebase please?
And also could you send a PR to the docs website?
Otherwise no one will ever know about this feature.

**Comment by dguo** [comment]
Sure, I can rebase tonight when I'm at my computer again.

I'm happy to send a PR for docs, but I didn't see an existing section on
the website for the offline mirror feature. Could you point me to the right
place?

On Mar 7, 2017 10:16 AM, "Konstantin Raev" <notifications@github.com> wrote:

@dguo <https://github.com/dguo>, could you rebase please?
And also could you send a PR to the docs website?
Otherwise no one will ever know about this feature.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/yarnpkg/yarn/pull/2836#issuecomment-284750142>, or mute
the thread

**Comment by bestander** [comment]
I think we only have a bog post
https://yarnpkg.com/blog/2016/11/24/offline-mirror/

Would you create a new section in the Docs please?
We'll convert the blog post into the doc after that.

**Comment by dguo** [comment]
Ok, it's rebased, the tests passed, and I submitted yarnpkg/website#409 for the docs.

**Comment by bestander** [comment]
Great job on pushing through the whole feature from RFC to great implementation with tests and docs, @dguo!

**Comment by dguo** [comment]
Thank you!

**Comment by rmacklin** [comment]
Wow, awesome! I'm participating in a hackathon today and this was actually the first thing I was gonna work on (I submitted the original issue), but I'm happy to see it's already done. Thanks @dguo!

Also, re: the tests and not having a test for `add`, I think there's an untested case we may want to cover. Suppose a team has been using yarn with an offline mirror for a while prior to the addition of this feature. Once they enable pruning, unless they had been manually pruning before, there will still be unused tarballs in the offline mirror. At that point, they should run `yarn` with no arguments (or equivalently `yarn install`) to prune all those unused tarballs, so that future add/remove/upgrade commands don't get mixed in with unrelated pruning.

So, I think it'd be good to add a test case for this scenario: start with a lockfile and a mirror that has more tarballs than what's in the lockfile, run `yarn` with no arguments (i.e. `yarn install`), verify that the unused tarballs were pruned. I've done that here: https://github.com/yarnpkg/yarn/pull/2880

We _could_ also add a similar test for running `add`, `remove`, and `upgrade` when the offline mirror already contains unused tarballs unrelated to the package that was added/removed/upgraded (we'd expect them to be pruned), though that may be overkill (if we don't think it's overkill, I can add those tests to the PR).

**Comment by rmacklin** [comment]
Is there any chance this could make it into a (pre-)release soon?

**Comment by bestander** [comment]
Yeah, we'll cut the branch mid next week

**Comment by rmacklin** [comment]
Awesome, and thanks for the speedy reply!

**Pull request merged**